pipelines/firebase/__init__.py

```diff
@@ -0,0 +1,84 @@
+"""Loads data from firebase realtime database"""
+
+from typing import Any, Iterator, List, Sequence
+
+import dlt
+from dlt.extract.source import DltResource
+from dlt.common.typing import TDataItem
+
+from firebase_admin import credentials, db
+
+
+@dlt.source(name="firebase")
+def firebase_source(
+    project_id: str =  dlt.secrets.value,
+    private_key_id: str =  dlt.secrets.value,
+    private_key: str = dlt.secrets.value,
+    client_email: str = dlt.secrets.value,
+    client_id: str = dlt.secrets.value,
+    database_url:  str = dlt.secrets.value
+) -> Sequence[DltResource]:
+    
+    def _get_data(credentials: Any, database_url: str) -> Iterator[TDataItem]:
+        """
+        Note: For this pipeline, we use a mock json template such below,
+        your firebase realtime data likely different:
+        {
+            "discography": [
+                {
+                    "albums": [
+                        {
+                            "album_name": str,
+                            "year": int
+                        }
+                    ],
+                    "band_name": str
+                }
+            ]
+        }
+        """
+
+        # initialize app
+        try:
+            firebase_admin.initialize_app(
+                credentials, 
+                {"databaseUrl": database_url}
+            )
+        except ValueError:
+            firebase_admin.get_app(name="[DEFAULT]")
+
+        ref = db.reference("/")
+        data = json.loads(ref.get())
+
+        """Please adjust the loop part according to your firebase database json schema"""
+        for item in data["discography"]:
```

Comment:
this should be placed outside of the pipeline. see https://dlthub.com/docs/general-usage/resource#filter-transform-and-pivot-data and demonstrate this transformer in demo script

Reply:
I have a lil bit confusion here, if we put transformation/parsing outside the pipeline, how the pipeline will store the parsed data? is there any example of pipeline that does it?